File: benchmarks/suite/vcall.cpp
#include <cstdio>
#include <cstdint>
struct Op { virtual int64_t apply(int64_t x)=0; virtual ~Op(){} };
struct AddOp:Op { int64_t k; AddOp(int64_t k):k(k){} int64_t apply(int64_t x)override{return x+k;} };
struct MulOp:Op { int64_t k; MulOp(int64_t k):k(k){} int64_t apply(int64_t x)override{return x*k;} };
int64_t run(Op*o,int64_t iters){int64_t acc=0;for(int64_t i=0;i<iters;i++)acc=o->apply(acc)%1000000;return acc;}
int main(){ AddOp a(3); MulOp m(7); printf("%ld\n",run(&a,50000000)+run(&m,50000000)); }

File: benchmarks/suite/vcall.rs
trait Op { fn apply(&self, x:i64)->i64; }
struct AddOp{k:i64} impl Op for AddOp { fn apply(&self,x:i64)->i64 { x+self.k } }
struct MulOp{k:i64} impl Op for MulOp { fn apply(&self,x:i64)->i64 { x*self.k } }
fn run(o:&dyn Op, iters:i64)->i64 { let mut acc=0i64; let mut i=0i64; while i<iters { acc=o.apply(acc)%1000000; i+=1; } acc }
fn main(){ let a=AddOp{k:3}; let m=MulOp{k:7}; println!("{}",run(&a,50000000)+run(&m,50000000)); }
